internal/provider/kubernetes/tlsroute.go

```diff
@@ -0,0 +1,323 @@
+	}
+
+	found := false
+	for i := range routeList.Items {
+		// See if this route from the list matched the reconciled route.
+		route := routeList.Items[i]
+		routeKey := utils.NamespacedName(&route)
+		if routeKey == request.NamespacedName {
+			found = true
+		}
+
+		// Store the tlsroute in the resource map.
+		r.resources.TLSRoutes.Store(routeKey, &route)
+		log.Info("added tlsroute to resource map")
+
+		// Get the route's namespace from the cache.
+		nsKey := types.NamespacedName{Name: route.Namespace}
+		ns := new(corev1.Namespace)
+		if err := r.client.Get(ctx, nsKey, ns); err != nil {
+			if errors.IsNotFound(err) {
+				// The route's namespace doesn't exist in the cache, so remove it from
+				// the namespace resource map if it exists.
+				if _, ok := r.resources.Namespaces.Load(nsKey.Name); ok {
+					r.resources.Namespaces.Delete(nsKey.Name)
+					log.Info("deleted namespace from resource map")
+				}
+			}
+			return reconcile.Result{}, fmt.Errorf("failed to get namespace %s", nsKey.Name)
+		}
+
+		// The route's namespace exists, so add it to the resource map.
+		r.resources.Namespaces.Store(nsKey.Name, ns)
+		log.Info("added namespace to resource map")
+
+		// Get the route's backendRefs from the cache. Note that a Service is the
+		// only supported kind.
+		for i := range route.Spec.Rules {
+			for j := range route.Spec.Rules[i].BackendRefs {
+				ref := route.Spec.Rules[i].BackendRefs[j]
+				if err := validateTLSRouteBackendRef(&ref); err != nil {
+					return reconcile.Result{}, fmt.Errorf("invalid backendRef: %w", err)
+				}
+
+				// The backendRef is valid, so get the referenced service from the cache.
+				svcKey := types.NamespacedName{Namespace: route.Namespace, Name: string(ref.Name)}
+				svc := new(corev1.Service)
+				if err := r.client.Get(ctx, svcKey, svc); err != nil {
+					if errors.IsNotFound(err) {
+						// The ref's service doesn't exist in the cache, so remove it from
+						// the resource map if it exists.
+						if _, ok := r.resources.Services.Load(svcKey); ok {
+							r.resources.Services.Delete(svcKey)
+							log.Info("deleted service from resource map")
+						}
+					}
+					return reconcile.Result{}, fmt.Errorf("failed to get service %s/%s",
+						svcKey.Namespace, svcKey.Name)
+				}
+
+				// The backendRef Service exists, so add it to the resource map.
+				r.resources.Services.Store(svcKey, svc)
+				log.Info("added service to resource map")
+			}
+		}
+	}
+
+	if !found {
+		// Delete the tlsroute from the resource map.
+		r.resources.TLSRoutes.Delete(request.NamespacedName)
+		log.Info("deleted tlsroute from resource map")
+
+		// Delete the Namespace and Service from the resource maps if no other
+		// routes exist in the namespace.
+		routeList = &gwapiv1a2.TLSRouteList{}
+		if err := r.client.List(ctx, routeList, &client.ListOptions{Namespace: request.Namespace}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error listing tlsroutes")
+		}
+		if len(routeList.Items) == 0 {
```

Comment:
is this safe ? what if HTTPRoute is accessing the same namespace and service ?

Reply:
if not ? I think it should be okay to handle this by raising a GH issue and fixing it in a follow up PR

Reply:
This should not be deleted, we'll have list httproutes too at this point to check whether any httproutes exist in the namespace. Same in httproute controller.
I'll fix this.